backend/api/src/jwkkeys.rs

```diff
@@ -0,0 +1,184 @@
+    pub expiration_time: Instant,
+}
+
+const FALLBACK_TIMEOUT: Duration = Duration::from_secs(60);
+
+pub async fn fetch_keys_for_config() -> anyhow::Result<JwkKeys> {
+    let http_response = reqwest::get(config::JWK_URL).await?;
+    let now = Instant::now();
+    let max_age = get_max_age(&http_response).unwrap_or(FALLBACK_TIMEOUT);
+    let expire_at = now - max_age;
+    let result = Result::Ok(http_response.json::<KeyResponse>().await?);
+
+    result.map(|res| JwkKeys {
+        keys: res.keys,
+        expiration_time: expire_at,
+    })
+}
+
+// Determines the max age of an HTTP response
+fn get_max_age(response: &Response) -> anyhow::Result<Duration> {
+    let headers = response.headers();
+    let header = headers.get("Cache-Control");
+    let header = header.ok_or_else(|| anyhow!("No Cache Control Header"))?;
+
+    parse_max_age_value(header.to_str()?)
+}
+
+fn parse_max_age_value(cache_control_value: &str) -> anyhow::Result<Duration> {
+    for token in cache_control_value.split(",") {
+        let mut key_value = token.split("=").map(str::trim);
+        let key = key_value
+            .next()
+            .expect("str split always gives at least one element");
+
+        if key.eq_ignore_ascii_case("max-age") {
+            let value = key_value
+                .next()
+                .ok_or_else(|| anyhow!("max-age is missing"))?;
+            return Ok(Duration::from_secs(value.parse()?));
+        }
+    }
+
+    bail!("No max-age specified");
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Claims {
+    pub sub: String,
+    pub iat: u64,
+    pub auth_time: u64,
+}
+
+fn get_current_timestamp() -> u64 {
+    let start = std::time::SystemTime::now();
+    start
+        .duration_since(std::time::UNIX_EPOCH)
+        .expect("Time went backwards")
+        .as_secs()
+}
+
+#[derive(Debug)]
+pub struct JwkVerifier {
+    keys: HashMap<String, JwkKey>,
+    issuer: String,
+    audience: String,
+}
+
+fn keys_to_map(keys: Vec<JwkKey>) -> HashMap<String, JwkKey> {
+    let mut keys_as_map = HashMap::new();
+    for key in keys {
+        keys_as_map.insert(key.kid.clone(), key);
+    }
+
+    keys_as_map
+}
+
+impl JwkVerifier {
+    pub fn new(keys: Vec<JwkKey>, issuer: String, audience: String) -> JwkVerifier {
+        JwkVerifier {
+            keys: keys_to_map(keys),
+            issuer,
+            audience,
+        }
+    }
+
+    pub fn verify(&self, token: &str) -> anyhow::Result<FirebaseId> {
+        let token_kid = jwt::decode_header(token)
+            .map_err(|e| anyhow!("error decoding jwt header: {}", e))?
+            .kid
+            .ok_or_else(|| anyhow!("Missing Key ID"))?;
+
+        let jwk_key = self
+            .keys
+            .get(&token_kid)
+            .ok_or_else(|| anyhow!("Invalid Key ID"))?;
+
+        let claims: Claims = self.decode_token_with_key(jwk_key, token)?.claims;
+        let now = get_current_timestamp();
+        if claims.auth_time > now || claims.iat > now {
```

Comment:
Just curious - is this enough of a claims check? I'm not totally familiar with the JWK workflow, so might be missing something - just kinda comparing to the Node implementation at https://github.com/firebase/firebase-admin-node/blob/24ac03fc9d36ee675719bf64cb1ff8cd7007f39b/src/auth/token-verifier.ts#L196 which seems to step through more of the claims (`iss`, `aud`, etc.)

Reply:
@izik1 explained via Slack, that this is dealt with in https://github.com/ji-devs/ji-cloud/pull/101/files#diff-c633486e93699bbc00467a0393ca9d2fR154 (and via the jsonwebtoken crate at https://docs.rs/jsonwebtoken/7.1.1/jsonwebtoken/struct.Validation.html